scripts/ci/tests/test_watchdog_logic.py

```
@@ -0,0 +1,147 @@
+        self.assertTrue(TEMPLATE.is_file(), f"Missing watchdog template: {TEMPLATE}")
+        template = TEMPLATE.read_text(encoding="utf-8")
+        marker = "        ZipFile: |\n"
+        self.assertIn(marker, template)
+        source_lines = []
+        for line in template[template.index(marker) + len(marker) :].splitlines(keepends=True):
+            if line.strip() and not line.startswith("          "):
+                break
+            source_lines.append(line)
+        source = textwrap.dedent("".join(source_lines))
+        namespace = {}
+        exec(compile(source, str(TEMPLATE), "exec"), namespace)
```

Ruff (0.16.1)

[error] Use of `exec` detected

(S102)